### ISSUE TITLE
Improve handling of ReflectionTypeLoadException in TypeCatalogue

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#1092]
Currently, if we get an error when scanning an assembly for extension points, we just emit a warning and dismiss the entire assembly. However, the error is often a `ReflectionTypeLoadException`, which gives us the option to be a little smarter:
- this exception exposes a `Types` property which is a list of types that were successfully loaded, so we could still scan those types (no need to dismiss the entire assembly just because one or two types couldn't be loaded)
- it also exposes a `LoaderExceptions` property with the errors that occurred when loading each type, so we could list the types that failed to load (and possibly indicate why they failed to load)

**Comment by thomaslevesque** [comment]
Example of possible output:

```
Warning: FakeItEasy failed to get some types from assembly 'DynamicProxyGenAssembly2, Version=0.0.0.0, Culture=neutral, PublicKeyToken=a621a9e7e5c32e69' while scanning for extension points. Some IArgumentValueFormatters, IDummyFactories, and IFakeOptionsBuilders in that assembly might not be available.
  Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
  The following loader exceptions occurred:
    - System.TypeLoadException: Could not load type 'Castle.Proxies.TypeWithNonConfigurableMethodProxy' from assembly 'DynamicProxyGenAssembly2, Version=0.0.0.0, Culture=neutral, PublicKeyToken=a621a9e7e5c32e69'.
    - System.TypeLoadException: Could not load type 'Castle.Proxies.Invocations.Object_Equals_10' from assembly 'DynamicProxyGenAssembly2, Version=0.0.0.0, Culture=neutral, PublicKeyToken=a621a9e7e5c32e69'.
```

**Comment by thomaslevesque** [comment]
> (and possibly indicate why they failed to load)

Actually, no. The `InnerException`s of the exceptions from `LoaderExceptions` seem to always be null.

**Comment by adamralph** [comment]
Would it be possible rename this issue to be more reflective of the functional benefit of this enhancement?

**Comment by thomaslevesque** [comment]
> Would it be possible rename this issue to be more reflective of the functional benefit of this enhancement?

Better now?

**Comment by adamralph** [comment]
👍 much better, thanks!

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 3.4.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/3.4.0).
